### PR TITLE
Fix #4513

### LINKED ISSUE #4513: Incorrect wrapping of proxy objects on every render
### How frequently does the bug occur?

All the time

### Description

https://github.com/realm/realm-js/blob/master/packages/realm-react/src/useObject.tsx#L78

```
    // Wrap object in a proxy to update the reference on rerender ( should only rerender when something has changed )
    return new Proxy(object, {});
```

The intention of this code is that the reference will only change if the object has changed to play nicely with reacts immutability conventions. However, the hook is not invoked only when the hook fires. It is invoked anytime the parent component renders. The component could be rendering because a totally unrelated change to some other component state which will cause unnecessary Proxy wrapping, unnecessarily new references and break React's immutability conventions.

Easier to explain with code

```
const initialObject = {};
function useObject() {
  return new Proxy(initialObject, {});
}

export default function App() {
  const [count, setCount] = useState(0);
  const object = useObject();
  useEffect(() => {
    // This effect will be invoked everytime the component
    // renders (eg click the button), even though the object
    // never actually changed
    console.log("object changed", object);
  }, [object]);

  return (
    <div className="App">
      <h1>{count}</h1>
      <button onClick={() => setCount((x) => x + 1)}>Click me</button>
    </div>
  );
}
```

https://codesandbox.io/s/modern-field-7ety98?file=/src/App.js:69-642

### Stacktrace & log output

_No response_

### Can you reproduce the bug?

Yes, always

### Reproduction Steps

_No response_

### Version

main

### What SDK flavour are you using?

Local Database only

### Are you using encryption?

No, not using encryption

### Platform OS and version(s)

ios

### Build environment

Which debugger for React Native: ..

### Cocoapods version

_No response_

### PR DESCRIPTION
**Pull request by mfbx9da4**
## What, How & Why?
Fixes https://github.com/realm/realm-js/issues/4513 and https://github.com/realm/realm-js/issues/4913

### CODE REVIEW

**Comment by mfbx9da4** [comment]
I didn't spend too long on this so might need some tweaking or cleaning up? @takameyer 

This won't fix https://github.com/realm/realm-js/issues/4518

This also won't fix the fact that:
 
> Using the same object via useObject in two different components will return different references 

(as mentioned in 4513). 

I imagine the code will have changed significantly to solve these latter two issues

**Comment by takameyer** [comment]
@mfbx9da4 Thanks for the PR!

Slightly different take than my idea:
```tsx
return function useObject<T>(type: string | { new (): T }, primaryKey: PrimaryKey): (T & Realm.Object) | null {
    const realm = useRealm();

    // Create a forceRerender function for the cachedObject to use as its updateCallback, so that
    // the cachedObject can force the component using this hook to re-render when a change occurs.
    const [version, forceRerender] = useReducer((x) => x + 1, 0);

    // Wrap the cachedObject in useMemo, so we only replace it with a new instance if `primaryKey` or `type` change
    const { object, tearDown } = useMemo(
      // TODO: There will be an upcoming breaking change that makes objectForPrimaryKey return null
      // When this is implemented, remove `?? null`
      () =>
        createCachedObject({
          object: realm.objectForPrimaryKey(type, primaryKey) ?? null,
          updateCallback: forceRerender,
        }),
      [type, realm, primaryKey],
    );

    // Invoke the tearDown of the cachedObject when useObject is unmounted
    useEffect(() => {
      return tearDown;
    }, [tearDown]);

    return useMemo(() => {
      // If the object has been deleted or doesn't exist for the given primary key, just return null
      if (object === null || object?.isValid() === false) {
        return null;
      } else {
        // Wrap object in a proxy to update the reference on rerender ( should only rerender when something has changed )
        return new Proxy(object, {});
      }
      // We need version as a dependency in order to trigger a new object reference
    }, [version, object]);
  };
```
We will need to write a test case for this and the other issues in order to avoid regressions.

**Comment by mfbx9da4** [comment]
Yeah cool! I like your way of doing it. It would be cleaner IMO if the cached object API was tweaked though

```
function useObject<T>(type: string | { new (): T }, primaryKey: PrimaryKey): (T & Realm.Object) | null {
    const realm = useRealm();

    const cachedObject = useMemo<T & Realm.Object | null>(() => createCachedObject(realm, type, primaryKey), [realm, type, primaryKey]);
    const [version, setVersion] = useState(0)
    

    useEffect(() => {
        return cachedObject?.subscribe(() => setVersion(x => x + 1));
    }, [cachedObject]);

    return useMemo(() => new Proxy(cachedObject.object, {}), [cachedObject, version]);
  };
```

**Review by takameyer — CHANGES_REQUESTED** [review]
Looks like we need to rebase and write a regression test.  If you got some time to do that, we would be very grateful 🙂 
Otherwise it may be some time before I get to it.

**Comment by takameyer** [comment]
Adding @tomduncalf and @kraenhansen to the review as well

File: packages/realm-react/src/useObject.tsx [hunk 1]
```diff
@@ -52,15 +52,26 @@ export function createUseObject(useRealm: () => Realm) {
...
     const { object, tearDown } = useMemo(
       // TODO: There will be an upcoming breaking change that makes objectForPrimaryKey return null
       // When this is implemented, remove `?? null`
       () =>
-        createCachedObject({
-          object: realm.objectForPrimaryKey(type, primaryKey) ?? null,
-          updateCallback: forceRerender,
-        }),
+        {
+          const updateCallback = () => {
```
Comment (sdandois) [inline packages/realm-react/src/useObject.tsx:63]: updateCallback is unused.

File: packages/realm-react/src/useObject.tsx [hunk 2]
```diff
@@ -74,7 +85,6 @@ export function createUseObject(useRealm: () => Realm) {
       return null;
     }
 
-    // Wrap object in a proxy to update the reference on rerender ( should only rerender when something has changed )
-    return new Proxy(object, {});
+    return objectRef.current
```
Comment (takameyer) [inline packages/realm-react/src/useObject.tsx:88]:
```suggestion
    // Return the current object reference, which should only cause a rerender when something has changed 
    return objectRef.current
```

File: packages/realm-react/src/useQuery.tsx [hunk 3]
```diff
@@ -65,8 +77,6 @@ export function createUseQuery(useRealm: () => Realm) {
       return tearDown;
     }, [tearDown]);
 
-    // This makes sure the collection has a different reference on a rerender
-    // Also we are ensuring the type returned is Realm.Results, as this is known in this context
-    return new Proxy(collection as Realm.Results<T & Realm.Object>, {});
+    return collectionRef.current;
```
Comment (takameyer) [inline packages/realm-react/src/useQuery.tsx:80]:
```suggestion
    // Return the current collection reference, which should only cause a rerender when something in the visible collection has changed 
    return collectionRef.current;
```

**Review by takameyer — CHANGES_REQUESTED** [review]
The code looks great, but I would like to see a regression test in both `useQueryRender` and `useObjectRender` that verifies that this fix works.  Functionally it should trigger a re-render on the parent component (using `useState`) and verify that the children were not re-rendered.

**Comment by takameyer** [comment]
I've squashed the current changes into a single commit and am continuing this PR on my own here:
https://github.com/realm/realm-js/pull/4958

**Pull request closed without merging**